diff --git a/dpdk b/dpdk
--- a/dpdk
+++ b/dpdk
@@ -1 +1 @@
-Subproject commit a8ba8524ac5581b424dedd46c599eeeb27a90073
+Subproject commit a8ba8524ac5581b424dedd46c599eeeb27a90073-dirty
diff --git a/include/spdk/nvme.h b/include/spdk/nvme.h
index 0252cfb5e..da25e50a1 100644
--- a/include/spdk/nvme.h
+++ b/include/spdk/nvme.h
@@ -958,6 +958,8 @@ int spdk_nvme_ctrlr_cmd_io_raw_with_md(struct spdk_nvme_ctrlr *ctrlr,
 int32_t spdk_nvme_qpair_process_completions(struct spdk_nvme_qpair *qpair,
 		uint32_t max_completions);
 
+int32_t nvme_pcie_qpair_available_completions(struct spdk_nvme_qpair *qpair);
+
 /**
  * Send the given admin command to the NVMe controller.
  *
diff --git a/lib/nvme/nvme_internal.h b/lib/nvme/nvme_internal.h
index 5804fdac8..278180760 100644
--- a/lib/nvme/nvme_internal.h
+++ b/lib/nvme/nvme_internal.h
@@ -976,6 +976,8 @@ struct spdk_nvme_ctrlr *spdk_nvme_get_ctrlr_by_trid_unsafe(
 	int nvme_ ## name ## _qpair_submit_request(struct spdk_nvme_qpair *qpair, struct nvme_request *req); \
 	int32_t nvme_ ## name ## _qpair_process_completions(struct spdk_nvme_qpair *qpair, uint32_t max_completions);
 
+int32_t nvme_pcie_qpair_available_completions(struct spdk_nvme_qpair *qpair);
+
 DECLARE_TRANSPORT(transport) /* generic transport dispatch functions */
 DECLARE_TRANSPORT(pcie)
 DECLARE_TRANSPORT(tcp)
diff --git a/lib/nvme/nvme_pcie.c b/lib/nvme/nvme_pcie.c
index 22699cfa1..69def42f4 100644
--- a/lib/nvme/nvme_pcie.c
+++ b/lib/nvme/nvme_pcie.c
@@ -2143,3 +2143,13 @@ nvme_pcie_qpair_process_completions(struct spdk_nvme_qpair *qpair, uint32_t max_
 
 	return num_completions;
 }
+
+int32_t
+nvme_pcie_qpair_available_completions(struct spdk_nvme_qpair *qpair)
+{
+	struct nvme_pcie_qpair	*pqpair = nvme_pcie_qpair(qpair);
+	struct spdk_nvme_cpl	*cpl;
+	cpl = &pqpair->cpl[pqpair->cq_head];
+	return cpl->status.p == pqpair->phase;
+}
+
