From 267039e65573a5bcb59e404464e0a7398e883b78 Mon Sep 17 00:00:00 2001
From: Josh Fried <joshuafried@gmail.com>
Date: Mon, 30 Dec 2024 17:27:30 -0600
Subject: [PATCH 1/4] config: extend max memseg lists

---
 config/rte_config.h | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/config/rte_config.h b/config/rte_config.h
index 86897de75e..aed406eea8 100644
--- a/config/rte_config.h
+++ b/config/rte_config.h
@@ -42,7 +42,7 @@
 #define RTE_CACHE_GUARD_LINES 1
 #define RTE_MAX_HEAPS 32
 #define RTE_MAX_LCORE_VAR 131072
-#define RTE_MAX_MEMSEG_LISTS 128
+#define RTE_MAX_MEMSEG_LISTS 1024
 #define RTE_MAX_MEMSEG_PER_LIST 8192
 #define RTE_MAX_MEM_MB_PER_LIST 32768
 #define RTE_MAX_MEMSEG_PER_TYPE 32768
-- 
2.43.0

